themes/default/content/blog/resource-methods-for-pulumi-packages/index.md

```diff
@@ -0,0 +1,609 @@
+    return err
+}
+```
+
+{{% /choosable %}}
+
+{{% /chooser %}}
+
+This post will show how you can provide resource methods from your [component packages]({{<relref "/docs/guides/pulumi-packages">}}).
+
+## Authoring Methods
+
+It's always been possible to provide module-level functions from Pulumi Packages. For example, the [`aws`]({{<relref "/docs/reference/pkg/aws">}}) package provides many module-level functions, such as the [`getAmi`]({{<relref "docs/reference/pkg/aws/ec2/getami">}}) function which can be used to get the ID of an existing Amazon Machine Image (AMI). Functions are declared in the package's [schema]({{<relref "/docs/guides/pulumi-packages/schema">}}) and their functionality is implemented in the provider through the `Invoke` RPC.
+
+Methods are authored in a similar manner to functions. Methods are declared in the schema and implemented in the provider's `Call` RPC (similar to `Invoke`).
+
+Let's walk through an example. We'll author a `Message` component which accepts a message as input and provides a `getMessage` method that accepts a recipient name and returns the message customized for the recipient. To get started authoring a component package, refer to the [package]({{<relref "/docs/guides/pulumi-packages">}}) documentation.
+
+### Schema
+
+We'll start with declaring the method and component in the [Pulumi schema]({{<relref "/docs/guides/pulumi-packages/schema">}}). First, define the function representing the method:
+
+```json
+  "functions": {
+    "example:index:Message/getMessage": {
+      "inputs": {
+        "properties": {
+          "__self__": {
+            "$ref": "#/resources/example:index:Message"
+          },
+          "recipient": {
+            "type": "string"
+          }
+        },
+        "required": ["__self__", "recipient"]
+      },
+      "outputs": {
+        "properties": {
+          "result": {
+            "type": "string"
+          }
+        },
+        "required": ["result"]
+      }
+    }
+  },
+```
+
+Our method has two required arguments: `__self__` and `recipient`. `__self__` is a special input required by all methods that represents the component resource and is typed as such. The method has one result named `result`.
+
+Next, define our `Message` component:
+
+```json
+  "resources": {
+    "example:index:Message": {
+      "isComponent": true,
+      "inputProperties": {
+        "message": {
+          "type": "string"
+        },
+      },
+      "requiredInputs": ["message"],
+      "properties": {
+        "message": {
+          "type": "string"
+        },
+      },
+      "required": ["message"],
+      "methods": {
+        "getMessage": "example:index:Message/getMessage"
+      }
+    }
+  },
+```
+
+Our component has a single required input/output property: `message`. The method is specified in the `methods` property, which references the method's function definition: `"getMessage": "example:index:Message/getMessage"`.
+
+Altogether our schema looks like:
```

Comment:
Updating language to improve accessibility (some folks will be on screen readers).

```suggestion
Here's our schema all together:
```